packages/zen-bull-vault/src/EmergencyWithdraw.sol

```diff
@@ -0,0 +1,255 @@
+        external
+    {
+        uint256 crabToRedeem = _zenBullAmount.wdiv(zenBullTotalSupplyForCrabWithdrawal).wmul(
+            IZenBullStrategy(zenBull).getCrabBalance()
+        );
+        (, uint256 wPowerPerpInCrab) = IZenBullStrategy(zenBull).getCrabVaultDetails();
+        uint256 wPowerPerpToRedeem =
+            crabToRedeem.wmul(wPowerPerpInCrab).wdiv(IERC20(crab).totalSupply());
+
+        zenBullTotalSupplyForCrabWithdrawal =
+            zenBullTotalSupplyForCrabWithdrawal.sub(_zenBullAmount);
+
+        IERC20(zenBull).transferFrom(msg.sender, address(this), _zenBullAmount);
+        _mint(msg.sender, _zenBullAmount);
+
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            3000,
+            wPowerPerpToRedeem,
+            _maxEthForWPowerPerp,
+            uint8(FLASH_SOURCE.EMERGENCY_WITHDRAW_ETH_FROM_CRAB),
+            abi.encodePacked(crabToRedeem, wPowerPerpToRedeem)
+        );
+
+        uint256 payout = IERC20(weth).balanceOf(address(this));
+        IWETH9(weth).withdraw(payout);
+        payable(msg.sender).sendValue(payout);
+
+        emit Withdraw(
+            msg.sender, _zenBullAmount, crabToRedeem, wPowerPerpToRedeem, payout, _zenBullAmount
+        );
+    }
+
+    /**
+     * @notice repay a portion of Euler debt and withdraw WETH collateral to this contract based on ZenBullEulerRecovery amount
+     * @dev this will revert of WETH to withdraw from Euler is greater than max, or limit price breach the tolerance %
+     * @dev if all ZenBullEulerRecovery are burnt, the ETH withdrawl will be activated
+     * @param _ratio ratio of WETH to withdraw from total WETH in Euler in 18 decimals (e.g 2e17 => 0.2 => 20%)
+     * @param _maxEthForUsdc max ETH to pay for 1 USDC
+     * @param _poolFee ETH/USDC Uni v3 pool fee
+     */
+    function emergencyRepayEulerDebt(uint256 _ratio, uint256 _maxEthForUsdc, uint24 _poolFee)
+        external
+    {
+        uint256 usdcToRepay = _ratio.wmul(IEulerDToken(dToken).balanceOf(zenBull));
+        uint256 wethToWithdraw = _ratio.wmul(IEulerEToken(eToken).balanceOfUnderlying(zenBull));
+
+        require(
+            wethToWithdraw <= MAX_WETH_PER_DEBT_REPAY,
+            "WETH to withdraw is greater than max per repay"
+        );
+
+        uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, usdc, weth, 420, false);
+
+        require(
+            _maxEthForUsdc <= ethUsdPrice.wmul((ONE.add(LIMIT_PRICE_TOLERANCE))),
+            "Max ETH for USDC greater than max price tolerance"
+        );
+
+        _exactOutFlashSwap(
+            weth,
+            usdc,
+            _poolFee,
+            usdcToRepay,
+            usdcToRepay.mul(WETH_DECIMALS_DIFF).wmul(_maxEthForUsdc),
+            uint8(FLASH_SOURCE.EMERGENCY_REPAY_EULER_DEBT),
+            abi.encodePacked(usdcToRepay, wethToWithdraw)
+        );
+
+        IWETH9(weth).withdraw(IERC20(weth).balanceOf(address(this)));
```

Comment:
When we accumulate eth here can't it be fully withdrawn by anyone calling emergencyWithdrawEthFromCrab?